### PR TITLE
Migrate from BytesRef to String

### PR DESCRIPTION
**Pull request by mfussenegger**
Using `BytesRef` instead of `String` was an optimization we did to avoid unnecessary utf8 to utf16 conversions. But we never benchmarked how much of an optimization it really is. 

In order to work with utf8-encoded BytesRef everywhere we patched ES to handle BytesRef within the `XContent` logic. In ES 6.3 they separated x-content into a separate package that has no Lucene dependency. So we either would've had to extend our patch or see if dropping it altogether is an option.

Below are a couple of benchmark runs. Note that these tend to fluctuate.

(V1 is master, V2 is this branch/PR)

```
Q: select * from uservisits limit 10
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        1.206 ±    2.054 |      0.686 |      1.037 |      1.278 |     64.834 |
|   V2    |        1.197 ±    2.292 |      0.649 |      1.014 |      1.273 |     72.333 |
mean:   -   0.80%

median: -   2.24%
Likely NOT significant

Q: select * from uservisits limit 100
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        2.403 ±    0.649 |      1.876 |      2.178 |      2.438 |      9.315 |
|   V2    |        2.661 ±    0.863 |      2.004 |      2.335 |      2.846 |     15.324 |
mean:   +  10.20%
median: +   6.95%
Likely significant

Q: select * from uservisits limit 100
C: 20
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        4.097 ±    2.474 |      1.969 |      3.466 |      4.559 |     24.737 |
|   V2    |        3.875 ±    1.397 |      2.108 |      3.628 |      4.081 |     12.708 |
mean:   -   5.58%
median: +   4.57%
Likely significant

Q: select * from uservisits limit 1000
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |       34.389 ±   10.645 |     17.699 |     34.074 |     45.650 |     61.440 |
|   V2    |       32.374 ±   11.228 |     18.907 |     30.185 |     42.054 |     65.195 |
mean:   -   6.03%
median: -  12.11%
Likely significant

Q: select * from uservisits limit 10000
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |      217.918 ±   29.409 |    174.948 |    205.586 |    242.067 |    337.787 |
|   V2    |      227.004 ±   27.582 |    184.960 |    213.629 |    247.225 |    325.161 |
mean:   +   4.08%
median: +   3.84%
Likely significant

Q: select * from uservisits limit 100000
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |     1902.934 ±   71.418 |   1790.907 |   1880.454 |   1971.413 |   2057.977 |
|   V2    |     2006.884 ±   72.106 |   1905.334 |   1984.902 |   2067.964 |   2199.732 |
mean:   +   5.32%
median: +   5.40%
Likely significant

Q: select * from uservisits order by "adRevenue" limit 10
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |       18.325 ±    3.158 |     16.375 |     16.936 |     18.310 |     72.836 |
|   V2    |       18.945 ±    3.634 |     16.436 |     17.057 |     21.893 |     84.801 |
mean:   +   3.33%
median: +   0.71%
Likely significant

Q: select * from uservisits order by "adRevenue" limit 10
C: 20
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |      130.491 ±   48.925 |     24.196 |    126.591 |    165.706 |    298.094 |
|   V2    |      130.390 ±   48.989 |     33.661 |    126.863 |    163.562 |    311.146 |
mean:   -   0.08%
median: +   0.21%
Likely NOT significant

Q: select "cCode", count(*) from uservisits group by 1 order by 2 desc limit 20
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |       40.700 ±    7.715 |     38.414 |     39.209 |     40.286 |    199.717 |
|   V2    |       38.858 ±    8.540 |     36.127 |     37.245 |     38.659 |    214.201 |
mean:   -   4.63%
median: -   5.14%
Likely significant

Q: select extract(day from "visitDate"), count(*) from uservisits group by 1 order by 2 desc limit 20
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |      136.368 ±   18.709 |     84.550 |    141.014 |    150.086 |    187.170 |
|   V2    |      136.273 ±   18.862 |     83.391 |    139.936 |    150.188 |    182.765 |
mean:   -   0.07%
median: -   0.77%
Likely NOT significant
```

```
Q: select * from uservisits where "sourceIP" = '25.193.131.52'
C: 1
| Version |         Mean Â±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        0.647 Â±    0.249 |      0.233 |      0.725 |      0.804 |      2.377 |
|   V2    |        0.638 Â±    0.300 |      0.232 |      0.714 |      0.794 |      6.238 |
mean:   -   1.41%
median: -   1.47%
Likely NOT significant

Q: select * from uservisits where
"sourceIP" = '162.122.2.75' or
"sourceIP" = '25.193.131.52' or
"sourceIP" = '138.43.129.246' or
"sourceIP" = '180.121.127.93' or
"sourceIP" = '55.104.52.130' or
"sourceIP" = '97.146.166.179' or
"sourceIP" = '166.53.95.120' or
"sourceIP" = '211.65.2.20' or
"sourceIP" = '40.65.238.81' or
"sourceIP" = '113.247.109.52' or
"sourceIP" = '100.8.41.251' or
"sourceIP" = '99.204.192.162' or
"sourceIP" = '182.165.186.253' or
"sourceIP" = '180.199.220.52' or
"sourceIP" = '38.9.148.224' or
"sourceIP" = '251.127.173.45' or
"sourceIP" = '214.48.17.113' or
"sourceIP" = '174.208.34.96' or
"sourceIP" = '49.53.176.22' or
"sourceIP" = '5.189.189.64'
C: 1
| Version |         Mean Â±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        1.351 Â±    0.427 |      1.163 |      1.319 |      1.362 |     14.525 |
|   V2    |        1.374 Â±    0.604 |      1.134 |      1.332 |      1.384 |     18.288 |
mean:   +   1.66%
median: +   0.94%
Likely NOT significant

Q: select * from uservisits order by "cCode" limit 10 offset 1000000
C: 1
| Version |         Mean Â±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |     1348.037 Â±  659.264 |   1079.886 |   1145.962 |   1209.898 |   3219.182 |
|   V2    |     1543.991 Â±  277.699 |   1262.711 |   1459.543 |   1876.402 |   1974.643 |
mean:   +  13.55%
median: +  24.07%
Likely NOT significant
```

```
Q: select * from uservisits where "sourceIP" = '25.193.131.52'
C: 1
| Version |         Mean Â±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        0.568 Â±    0.201 |      0.211 |      0.626 |      0.697 |      1.657 |
|   V2    |        0.763 Â±    0.144 |      0.540 |      0.751 |      0.795 |      2.808 |
mean:   +  29.32%
median: +  18.15%
Likely significant

Q: select * from uservisits where
"sourceIP" = '162.122.2.75' or
"sourceIP" = '25.193.131.52' or
"sourceIP" = '138.43.129.246' or
"sourceIP" = '180.121.127.93' or
"sourceIP" = '55.104.52.130' or
"sourceIP" = '97.146.166.179' or
"sourceIP" = '166.53.95.120' or
"sourceIP" = '211.65.2.20' or
"sourceIP" = '40.65.238.81' or
"sourceIP" = '113.247.109.52' or
"sourceIP" = '100.8.41.251' or
"sourceIP" = '99.204.192.162' or
"sourceIP" = '182.165.186.253' or
"sourceIP" = '180.199.220.52' or
"sourceIP" = '38.9.148.224' or
"sourceIP" = '251.127.173.45' or
"sourceIP" = '214.48.17.113' or
"sourceIP" = '174.208.34.96' or
"sourceIP" = '49.53.176.22' or
"sourceIP" = '5.189.189.64'
C: 1
| Version |         Mean Â±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        1.365 Â±    0.644 |      1.149 |      1.324 |      1.376 |     16.045 |
|   V2    |        1.330 Â±    0.109 |      1.133 |      1.307 |      1.370 |      2.660 |
mean:   -   2.63%
median: -   1.28%
Likely NOT significant

Q: select * from uservisits order by "cCode" limit 10 offset 1000000
C: 1
| Version |         Mean Â±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |     1480.973 Â±  522.365 |   1134.376 |   1173.058 |   1981.174 |   2567.611 |
|   V2    |     1279.188 Â±  113.122 |   1181.517 |   1234.243 |   1336.633 |   1561.262 |
mean:   -  14.62%
median: +   5.08%
Likely NOT significant
```

```
Q: select count(*) from (select distinct x from t090) t
C: 1
| Version |         Mean Â±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |      747.942 Â±   76.172 |    664.588 |    736.473 |    741.999 |   1015.528 |
|   V2    |      651.543 Â±   31.712 |    612.774 |    645.238 |    655.768 |    802.843 |
mean:   -  13.78%
median: -  13.21%
Likely significant

Q: select count(*) from (select distinct x from t075) t
C: 1
| Version |         Mean Â±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |      643.728 Â±   19.505 |    616.492 |    645.748 |    651.108 |    805.049 |
|   V2    |      599.472 Â±   17.553 |    562.823 |    598.206 |    600.839 |    699.020 |
mean:   -   7.12%
median: -   7.64%
Likely significant

Q: select count(*) from (select distinct x from t050) t
C: 1
| Version |         Mean Â±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |      564.506 Â±   57.194 |    525.403 |    551.539 |    556.223 |    867.346 |
|   V2    |      568.261 Â±   36.707 |    541.920 |    562.736 |    566.096 |    848.431 |
mean:   +   0.66%
median: +   2.01%
Likely NOT significant

Q: select count(*) from (select distinct x from t025) t
C: 1
| Version |         Mean Â±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |      501.178 Â±    7.648 |    493.269 |    499.288 |    501.595 |    542.683 |
|   V2    |      498.642 Â±   24.160 |    463.427 |    494.107 |    505.254 |    738.254 |
mean:   -   0.51%
median: -   1.04%
Likely NOT significant
```

- [x] User relevant changes are recorded in ``CHANGES.txt``
 - [x] Touched code is covered by tests
 - [x] Documentation has been updated if necessary
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed

### CODE REVIEW

**Comment by chaudum** [comment]
So from looking at the benchmark results, the conclusion is that the change is likely not significant (neither way), right? :+1:

**Comment by mfussenegger** [comment]
> So from looking at the benchmark results, the conclusion is that the change is likely not significant

Yep I'd also interpret it that way. At worst it's a slight performance decrease, but I think the removal of complexity would justify it.

**Comment by chaudum** [comment]
> > So from looking at the benchmark results, the conclusion is that the change is likely not significant
> 
> Yep I'd also interpret it that way. At worst it's a slight performance decrease, but I think the removal of complexity would justify it.

Definitely!

**Comment by mfussenegger** [comment]
retest this please

**Comment by mfussenegger** [comment]
retest this please

File: sql/src/main/java/io/crate/execution/engine/collect/GroupByOptimizedIterator.java [hunk 1]
```diff
@@ -236,7 +237,7 @@ static boolean hasHighCardinalityRatio(Supplier<Engine.Searcher> acquireSearcher
 
                 @Override
                 public Row apply(Map.Entry<BytesRef, Object[]> entry) {
-                    cells[0] = entry.getKey();
+                    cells[0] = BytesRefs.toString(entry.getKey());
```
Comment (chaudum) [inline sql/src/main/java/io/crate/execution/engine/collect/GroupByOptimizedIterator.java]: I'm uncertain, why we still have BytesRef here.

File: sql/src/main/java/io/crate/expression/scalar/SubstrFunction.java [hunk 2]
```diff
@@ -74,80 +71,37 @@ public BytesRef evaluate(Input[] args) {
...
-            }
-
-            // Anything not covered above is invalid UTF8.
-            throw new IllegalArgumentException("substr: invalid UTF8 string found.");
-        }
-
-        // Check if we didn't go over the limit on the last character.
-        if (pos > limit) throw new IllegalArgumentException("begin index must not be > end index");
-        return new BytesRef(bytes, posBegin, pos - posBegin);
+    static String substring(String value, int begin, int end) {
```
Comment (chaudum) [inline sql/src/main/java/io/crate/expression/scalar/SubstrFunction.java]: This is definitely my favorite change :)

**Comment by mfussenegger** [comment]
retest this please

**Comment by mfussenegger** [comment]
retest this please

File: sql/src/main/java/io/crate/execution/engine/collect/GroupByOptimizedIterator.java (hunk 1 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/execution/engine/collect/GroupByOptimizedIterator.java]: We keep using BytesRef in the map in order to do cheap exists checks - so we don't have to create a String for each document to check if we've seen it already.

**Comment by mfussenegger** [comment]
retest this please

**Comment by mfussenegger** [comment]
retest this please

**Pull request merged**